File: Marcianitos/src/deepspace/Damage.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author laura
 */
public class Damage {
    private int nShields;
    private int nWeapons;
    private ArrayList<WeaponType> weapons; 
    
    static final int NO_USO = -1;
    
    Damage(int w, int s){
        this.nWeapons = w;
        this.nShields = s;
        weapons = null;
    }
    
    Damage(ArrayList<WeaponType> wl, int s) {
        this.weapons = wl;
        this.nShields = s;
        this.nWeapons = NO_USO; 
    }
    
    Damage (Damage d) {
        this.nShields = d.getNShields();
        this.nWeapons = d.getNWeapons();
        this.weapons = d.getWeapons();
    }
    
    DamageToUI getUIversion() {
        return new DamageToUI(this);
    }
    
    private int arrayContainsType(ArrayList<Weapon> w, WeaponType t) {        
        int pos = -1;
        boolean encontrado = false;
        int i = 0;
        
        while(i<w.size() && !encontrado) {
            if (w.get(i).getType() == t) {
                encontrado = true;
                pos = i;
            }
            else
                i++;
        }
        
        return pos;
    }
    
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) { // funciona 
        
        Damage ajustado;
        int l_nshields = Math.min(s.size(), this.getNShields());

        if (this.getWeapons() == null) { // estamos usando solo numero de weapon
            int l_nweapons = Math.min(w.size(), this.getNWeapons());
            ajustado = new Damage(l_nweapons, l_nshields);
        }
        else {  // estamos usando array de weapontype
            ArrayList<WeaponType> result = new ArrayList();
            ArrayList<Weapon> aux = new ArrayList(w);
            
                    
            for (int i = 0; i<this.getWeapons().size(); i++) {
                WeaponType element = this.getWeapons().get(i);
                int indice = this.arrayContainsType(aux, element);
                if (indice != -1) {
                    result.add(element);
                    aux.remove(indice);
                }
            }            
            ajustado = new Damage(result, l_nshields);
        }

        return ajustado;
    }
    
    public void discardWeapon(Weapon w) {   // funciona
        if (getWeapons() != null) 
            getWeapons().remove(w.getType());
        else
            if (getNWeapons() > 0)
                this.nWeapons--;
    }
    
    public void discardShieldBooster() { //funciona
        if (getNShields() > 0)
            this.nShields --;
    }
    
    public boolean hasNoEffect(){   //funciona
        boolean no_weapon_damage = nWeapons == NO_USO ? (this.weapons.isEmpty()) : (this.nWeapons == 0);

        return (no_weapon_damage && (this.nShields == 0));

    }
    
    public int getNShields() {
        return this.nShields;
    }
    
    public int getNWeapons() {
        return this.nWeapons;
    }
    
    public ArrayList<WeaponType> getWeapons(){
        return this.weapons;
    }
    
    public String toString() {
    
        String out = "numero shields: " + nShields + "\n";
        if (getWeapons() != null) {
            out += getWeapons().toString();
        }
        else {
            out += "numero weapons: " + nWeapons + "\n";
        }
        
        return out;
    }
}

File: RUBY/src/Damage.rb
require_relative 'DamageToUI'
require_relative 'WeaponType'
require_relative 'Weapon'

module Deepspace
	class Damage
		
		@@NOUSO=-1
		
		def initialize(s,w,wp)
			@nShields=s
			@nWeapons=w
			@weapons=wp
		end
		
		def self.newNumericWeapons(w, s)
			return new(s, w, nil)
		end
		
		def self.newSpecificWeapons(wl, s)
			return new(s, @@NOUSO, wl)
		end
		
		def self.newCopy(d)
			if d.nWeapons == @@NOUSO
				return newSpecificWeapons(d.weapons, d.nShields)
			else
				return newNumericWeapons(d.nWeapons, d.nShields)
			end
		end
		
		def getUIversion()
			return damegeToUI.new(self)
		end
		
		attr_reader :nShields, :nWeapons, :weapons
		
		def adjust(w,s)   ## no estoy muy seguro
		
			l_nshields = [s.length, nShields].min

			if weapons==nil
			
				l_nweapons = [w.length, nWeapons].min
				
				return Damage.newNumericWeapons(l_nweapons,l_nshields)

			else
				result = []
				w_aux = w.clone
				
				weapons.each do |element|
					indice = arrayContainsType(w_aux, element)
              
					if indice != -1
						result.push(element)
						w_aux.delete_at(indice)
					end
				end

				Damage.newSpecificWeapons(result, l_nshields)
			end
		end
		
		
		def discardWeapon(w)
		
			if @nWeapons == @@NOUSO
			
				if @weapons.length != 0
					
					indice = @weapons.index(w.type)
					
					if indice != nil
						
						@weapons.delete_at(indice)
					end
				end
				
			elsif
				
				if @nWeapons > 0
					@nWeapons -= 1
				end
			end
			
		end
			
			
		def discardShieldBooster()
					
			if @nShields > 0
				
				@nShields -= 1
				
			end
			
		end
		
		def hasNoEffect()
		
			if @nWeapons == @@NOUSO
			
				retorno = @weapons.length + @nShields == 0
				
			else
			
				retorno = @nShields + @nWeapons == 0
				
			end
			
			return retorno
		end
		
		def to_s
			out="Damage - nShields: #{@nShields}"
						
			if (nWeapons != @@NOUSO)
				out+=", nWeapons: #{@nWeapons}\n"
			else
				out+="\tWeapons: [#{@weapons.join(', ')}]\n"
			end
			out+="------- end of Damage -------"
			return out
		end
		
		private 
		def arrayContainsType(w,t)
			
			contador=0
			retorno = -1
			
			w.each do |weapon_aux|
			
				if weapon_aux.type == t
					retorno = contador
				else
					contador+=1
				end
			end
			
			return retorno					
		end
		
	end
end			
